src/librbd/io/Dispatcher.h

```diff
@@ -101,59 +133,81 @@ class Dispatcher : public DispatchInterfaceT {
 
     auto dispatch_layer = dispatch_spec->dispatch_layer;
 
+    // If this is the first time into the send function then record this in the overall op_tracker
+    if (dispatch_spec->dispatch_result == DISPATCH_RESULT_INIT) {
+      m_op_tracker->start_op();
+    }
+    if (dispatch_spec->dispatch_result == DISPATCH_RESULT_REFRESH) {
+      ldout(cct, 20) << "rstart. dispatch_spec=" << dispatch_spec << dendl;
+      m_op_tracker->start_op();
+    }
+
     // apply the IO request to all layers -- this method will be re-invoked
     // by the dispatch layer if continuing / restarting the IO
     while (true) {
-      m_lock.lock_shared();
       dispatch_layer = dispatch_spec->dispatch_layer;
       auto it = m_dispatches.upper_bound(dispatch_layer);
       if (it == m_dispatches.end()) {
         // the request is complete if handled by all layers
         dispatch_spec->dispatch_result = DISPATCH_RESULT_COMPLETE;
-        m_lock.unlock_shared();
         break;
       }
 
       auto& dispatch_meta = it->second;
       auto dispatch = dispatch_meta.dispatch;
-      auto async_op_tracker = dispatch_meta.async_op_tracker;
       dispatch_spec->dispatch_result = DISPATCH_RESULT_INVALID;
 
-      // prevent recursive locking back into the dispatcher while handling IO
-      async_op_tracker->start_op();
-      m_lock.unlock_shared();
-
       // advance to next layer in case we skip or continue
       dispatch_spec->dispatch_layer = dispatch->get_dispatch_layer();
 
       bool handled = send_dispatch(dispatch, dispatch_spec);
-      async_op_tracker->finish_op();
+      // This will end up in ImageDispatchSpec::C_Dispatcher::complete() where it will look at the dispatch_spec result.
+      // If it is CONTINUE then it will call back into this function (for the next layer)
+      // If it is COMPLETE then it will call finish - in ImageDispatchSpec this will just delete the dispatchSpec
+
+      // Alternatively this will end up in ObjectDispatchSpec::C_Dispatcher::complete() where it will look at the dispatch_spec result.
+      // If it is CONTINUE then it will call back into this function (for the next layer)
+      // If it is COMPLETE then it will call finish - this will call the complete() method (and then delete the object dispatchSpec)
+      //                                              the complete() method will call AioCompletion::complete which in turn will call AioCompletion::notify_callbacks_complete()
```

Comment:
Minor nitpick but AioCompletion::notify_callbacks_complete() is now called AioCompletion::mark_complete_and_notify().